mediaplatform/models.py

```diff
@@ -0,0 +1,146 @@
+import uuid
+
+from django.db import models
+import django.contrib.postgres.fields as pgfields
+
+
+class Permission(models.Model):
+    """
+    Specify whether a user has permission to perform some action.
+
+    A user has permission to perform the action if *any* of the following are true:
+
+    * They have a crsid and that crsid appears in :py:attr:`~.crsids`
+    * The numeric id of a lookup group which they are a member of appears in
+      :py:attr:`~.lookup_groups`
+    * The instid of an an institution they are a member of appears in
+      :py:attr:`~.lookup_insts`
+    * The :py:attr:`~.is_public` flag is ``True``
+    * The user is not anonymous and :py:attr:`is_signed_in` is ``True``
+
+    """
+    #: Primary key
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    #: List of crsids of users with this permission
+    crsids = pgfields.ArrayField(models.TextField(), blank=True, default=[])
+
+    #: List of lookup groups which users with this permission belong to
+    lookup_groups = pgfields.ArrayField(models.BigIntegerField(), blank=True, default=[])
+
+    #: List of lookup institutions which users with this permission belong to
+    lookup_insts = pgfields.ArrayField(models.TextField(), blank=True, default=[])
+
+    #: Do all users (including anonymous ones) have this permission?
+    is_public = models.BooleanField(default=False)
+
+    #: Do all signed in (non-anonymous) users have this permission?
+    is_signed_in = models.BooleanField(default=False)
+
+
+class MediaItem(models.Model):
+    """
+    An individual media item in the media platform.
+
+    """
+    #: Primary key
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```

Comment:
This should be shorter to reference it in the url or are you thinking in doing it at runtime?

Reply:
I think the URL encoding is a separate problem. Base64 encoding the UUIDs should lead to URLs which look like https://alpha.media.cam.ac.uk/media/IeGheechaegool9ahth7ee which I think is probably OK. We could use a random 64-bit integer instead which would lead to URLs like https://alpha.media.cam.ac.uk/media/ool2ieFe0e. I'm neutral on which we choose so I'm happy to do whatever you'd prefer.

Reply:
As a separate issue, I think it'd be worth running the Base64 encoded ids through, e.g., https://github.com/Volnus/YouTube-Blacklist to avoid accidentally having media with URLs containing bad words.

Reply:
> I'm neutral on which we choose so I'm happy to do whatever you'd prefer.

Actually, I'm coming down on the side of a random 64-bit integer as providing a nicer compromise so I think I'll re-work this to use random 64-bit integer ids. I'll work the Birthday Paradox probabilities first though.

Alternatively, we could use an approach like https://wiki.postgresql.org/wiki/Pseudo_encrypt or https://wiki.postgresql.org/wiki/Pseudo_encrypt_constrained_to_an_arbitrary_range but I think that's becoming needlessly complex.

Reply:
So, according to the table at https://en.wikipedia.org/wiki/Birthday_problem#Probability_table, to get a 1 in a million chance of collision with 64-bit ids we'd need around 6 million media items/collections. For 128-bit UUIDs, that becomes 2.6 x 10^16.

Further, collision only actually matters if we use 64-bit ids for more than one table (e.g. media items and collections). Within a table, Postgres saves us from collision. With that the probability that a) a collision occurs and b) it happens to be between a collection and a media item probably becomes tolerably small.